### PR TITLE
Network Layer and Model Refactor

### LINKED ISSUE #72: Poor performance due to missing caching
So from what I've tested the app does not seem to cache most of the data. Most noticebly it happend several times that the app was unresponsive due to load some data (10-20 seconds in the case of the calendar).

We will run into problems with the TUM. Especially the calendar uses a lot of resources on their mainframe. That should be updated once every 7 days or so at the very most! (Ask @xsrf for more details on this)

Please take a look at the Androids caching mechanics: https://github.com/TCA-Team/TumCampusApp/blob/master/app/src/main/java/de/tum/in/tumcampusapp/managers/CacheManager.java

### LINKED ISSUE #90: Refactor URLs
Refactor the creation of URLs by writing all the baseURLs in a JSON file . And finally get rid of all the warnings from Sonarqube

### LINKED ISSUE #109: Grades and Books should each be displayed in a single card
Todo: Create dedicated model for each of these cards

### LINKED ISSUE #156: Caching Everything
If individual room maps images aren't too large, we should consider caching them on device.

Obvious advantage of this is performance: if we don't have to load the images from the internet, we can display them way faster if the user is on a slow connection.

### PR DESCRIPTION
**Pull request by mathiasquintero**
This redoes the entire networking layer and removes SwiftyJSON and Alamofire from our dependencies.

It makes our networking a bit more type-safe and relies less on casting.

And it introduces caching

Fixes #72
Fixes #109 
Starts progress in #43 & #165
Fixes #156
Fixes #90

While I was at it I also fixed: #121 with the same workaround as android

### CODE REVIEW

**Comment by mammuth** [comment]
Yehii! 🎉 

What is your goal with this big change? Do you want this to be into this weeks release? 
I'd suggest to focus on the main issues we have (mainly login, calendar) and then bring in this major refactoring later after we had enough time reviewing and testing everything!

**Comment by mathiasquintero** [comment]
Managers remaining: 
- [ ] MVG
- [ ] Login Manager
- [ ] Location Provider (Refactor getting location for Cafeteria and MVG)

**Comment by TG908** [comment]
@mathiasquintero I'd would be great if the new model classes would contain a collection of DataElements. So I can for example display multiple BookRentals in one Collection View inside the card or multiple Grades in one CollectionView inside the card.

**Comment by mathiasquintero** [comment]
@TG908 that's already in. Check out the single manager protocol. It's a manager whose data can be reduced to a single element. For most it's just taking the first element. But for the other ones it can return any other data

**Comment by mathiasquintero** [comment]
@TG908 to see what I want to do with the extra models for the cards check the `MVGManager` here where I already created a `DetailedStation` model that includes a station and the close departure times. And this would be the DataElement for the card

**Comment by TG908** [comment]
> @TG908 to see what I want to do with the extra models for the cards check the MVGManager here where I already created a DetailedStation model that includes a station and the close departure times. And this would be the DataElement for the card:

looks great

**Comment by mathiasquintero** [comment]
@TG908 I still have a bug with setting the image in the news view. For some reason I'm setting it and it's not being shown. Could you please take a look?

**Comment by TG908** [comment]
@mathiasquintero on my device it is shown (john wick 2 movie poster)

**Comment by mathiasquintero** [comment]
@TG908 and @mammuth please check it out and test it manually. I moved a lot of code around and I'm pretty sure I got a bad merge in at some point...

**Comment by mathiasquintero** [comment]
The news is not sorted correctly. Good catch. And we're now taking the latest, which is always a future movie.
I'm not sure what's causing the long loading time. But I'll debug it

**Comment by TG908** [comment]
Doing the viewUpdate on the main queue reduce name loading time significantly.

**Comment by mathiasquintero** [comment]
Sweeft handles everything in the global background thread unless specified otherwise.

That’s why the UI code has to specify it want the success handler to be called in the main queue.

Other than that we don’t have any custom dispatching behaviors

**Comment by mathiasquintero** [comment]
However my design was that promises have their own completionqueues where they call all their completion handlers.

So the queue has to be known during initialization. Perhaps this design is flawed and we should be able to specify it for each handler. 

I added a hacky extension to avoid that in the extensions file by using map.

**Comment by TG908** [comment]
I am having issues with the library card

**Comment by mathiasquintero** [comment]
@TG908 Can you please be more specific?

**Comment by TG908** [comment]
I’m getting no Library Card and no success message on login.

**Comment by TG908** [comment]
@mathiasquintero the card doesn't show up and when trying to log in theres no success alert

**Comment by TG908** [comment]
I think i found the bug: `// TODO: Access BookRentalsManager`

**Comment by TG908** [comment]
@mathiasquintero Do you know if the search Manager can also search in categories like, people, rooms and lectures?

**Comment by mathiasquintero** [comment]
@TG908 Ok. I'm fixing it

**Comment by mathiasquintero** [comment]
@TG908 I don't unserstand your question. We don't have a single search manager. We have many that are concatenated whenever you perform a search.

These search for People, Rooms and Lectures. But we can always extend them. Just look for `TumDataManager.searchManagers`

**Comment by mathiasquintero** [comment]
@mammuth @TG908 any other bugs or concerns, because otherwise, I think I'm done with this one...
And if we see any more bugs later we can also fix them

**Comment by TG908** [comment]
I think we should create a new branch and start integration of all the changes.

**Comment by mathiasquintero** [comment]
@TG908 I'm not sure what you mean by integration?

**Comment by TG908** [comment]
> @TG908 I'm not sure what you mean by integration. This is all integrated. What we're missing is the new Cards and UIs we have planned

Did you change the search controller ?

File: TUM Campus App/ImageDownloader.swift [hunk 1]
```diff
@@ -6,54 +6,59 @@
 //  Copyright © 2015 LS1 TUM. All rights reserved.
 //
 
-import Foundation
 import UIKit
-import Alamofire
+import Sweeft
 
-class ImageDownloader {
+class ImageViewBinding {
```
[the remark this answers is not in the document]
Comment (mathiasquintero) [reply TUM Campus App/ImageDownloader.swift]: Nope. I need it to be a fixed reference Type. I could use NSObject but that's not clear enough about what I want to do.

**Pull request merged**